scipy/_lib/_util.py

```diff
@@ -337,3 +338,69 @@ def getargspec_no_self(func):
         if argspec.args[0] == 'self':
             argspec.args.pop(0)
         return argspec
+
+
+class PoolWrapper(object):
+    """
+    Wrapper for working with objects that have map-like methods, such as
+    `multiprocessing.Pool`. Used for parallelisation.
+
+    Parameters
+    ----------
+    pool : int or map-like object
+        If `pool` is an `int` then it specifies the number of threads to
+        use for parallelization. If `pool == 1`, then no parallel
+        processing is used and the map builtin is used. If `pool == 0` then
+        the pool will utilise all available CPU.
+        If pool is an object with a map method that follows the same
+        calling sequence as the built-in map function, then this pool is
+        used for parallelisation.
+    """
+    def __init__(self, pool=1):
+        self.pool = None
+        self._mapfunc = map
+        self._own_pool = False
+
+        if hasattr(pool, 'map'):
+            self.pool = pool
+            self._mapfunc = self.pool.map
+        else:
+            # user supplies a number
+            if int(pool) == 0:
+                # use as many processors as possible
+                self.pool = Pool()
```

Comment:
We should use `joblib.Parallel` here, it's superior to plain `multiprocessing` (better handling of numpy arrays, can be interrupted with Ctrl-C, probably some more relevant improvements). We can vendor it, so no additional dependency.

Reply:
1) AFAICT (and I might be wrong), the `multiprocessing.Pool` is now keyboard interruptible. My basis for that is [here](https://github.com/dfm/emcee/blob/master/emcee/interruptible_pool.py), but I don't know where to go to get the truth.
2) The rationale for being able to supply an object with a map method is that it's a common design, one can use: `map`, `Pool.map`, `mpi4py. MPIPoolExecutor.map`, etc. Unfortunately `joblib.Parallel` doesn't have a map method, so extra code has to be written to be able to use its `__call__` method (pretty sure it's doable).
3) If a `joblib.Parallel` is not used within a context manager every `__call__` [initialises](https://github.com/joblib/joblib/blob/master/joblib/parallel.py#L761) and [closes](https://github.com/joblib/joblib/blob/master/joblib/parallel.py#L816) its multiprocessing Pool. The design of the `DifferentialEvolutionSolver` (DES) object is that after initialisation one can advance (evolve) the object by advancing the `DES.next` generator. Each successive call to `DES.next` will involve repeated requests to `joblib.Parallel.__call__`, meaning for N advances of the generator one will have created and destroyed N `multiprocessing.Pool`s. This will add significant overhead (from `joblib` documentation):
> Calling Parallel several times in a loop is sub-optimal because it will create and destroy a pool of workers (threads or processes) several times which can cause a significant overhead.

Overhead would be reduced if the `joblib.Parallel` object could stay 'alive' for the entire existence of the `DES` object.

Reply:
```
from multiprocessing import Pool
a = np.linspace(0, 1000, 1000)
def f(a, p=None):
        if p is None:
            with Pool() as p:
                b = p.map(np.sin(a))
        else:
            b = p.map(np.sin, a)
        return b
```
```
>>> %timeit f(a)
122 ms ± 748 µs per loop (mean ± std. dev. of 7 runs, 10 loops each)
>>> p = Pool()
>>> %timeit f(a, p=p)
53.3 ms ± 637 µs per loop (mean ± std. dev. of 7 runs, 10 loops each)

Reply:
The memmaping of joblib is attractive, but it's only useful if the memmaped array is going to be part of the mapped function call. For optimisation problems I doubt that the parameter array will ever be large enough to make that worthwhile.
For other problems in the scipy stable it would probably make sense to have some kind of joblib functionality.

Reply:
Another option may be to use `ProcessPoolExecutor` instead of a plain `Pool`